### PR TITLE
[SPARK-25993][SQL][TESTS] Add test cases for CREATE EXTERNAL TABLE with subdirectories

### PR DESCRIPTION
**Pull request by kevinyu98**
### What changes were proposed in this pull request?

This PR aims to add these test cases for resolution of ORC table location reported by [SPARK-25993](https://issues.apache.org/jira/browse/SPARK-25993)
also add corresponding test cases for Parquet table.

### Why are the changes needed?

The current behavior is complex, this test case suites are designed to prevent the accidental behavior change. This pr is rebased on master, the original pr is [23108](https://github.com/apache/spark/pull/23108)

### Does this PR introduce any user-facing change?

No. This adds test cases only.

### How was this patch tested?

This is a new test case.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 1]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+              s"""
+                 |CREATE EXTERNAL TABLE tbl1(
+                 |  c1 int,
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}"}'""".stripMargin
+            sql(topDirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("select * from tbl1"), Nil)
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Shall we capitalize the SQL statement like `SELECT * FROM tbl1`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 2]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+            } else {
+              intercept[IOException](sql("select * from tbl1").show())
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: We had better check the exception message.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 3]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+              s"""
+                 |CREATE EXTERNAL TABLE tbl2(
+                 |  c1 int,
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}/l1/"}'""".stripMargin
+            sql(l1DirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("select * from tbl2"),
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: `SELECT * FROM tbl2`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 4]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                (1 to 2).map(i => Row(i, i, s"parq$i")))
+            } else {
+              intercept[IOException](sql("select * from tbl2").show())
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Please check the exception message.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 5]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+              s"""
+                 |CREATE EXTERNAL TABLE tbl3(
+                 |  c1 int,
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}/l1/l2/"}'""".stripMargin
+            sql(l2DirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("select * from tbl3"),
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: `SELECT * FROM tbl3`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 6]
```diff
@@ -222,4 +223,115 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                (3 to 4).map(i => Row(i, i, s"parq$i")))
+            } else {
+              intercept[IOException](sql("select * from tbl3").show())
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Please check the exception message.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 7]
```diff
@@ -170,4 +170,155 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
     testMergeSchemasInParallel(OrcFileOperator.readOrcSchemasInParallel)
   }
+
+  test("SPARK-25993 CREATE EXTERNAL TABLE with subdirectories") {
+    Seq(true, false).foreach { convertMetastore =>
+      withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> s"$convertMetastore") {
+        withTempDir { dir =>
+          try {
+            hiveClient.runSqlHive("USE default")
+            hiveClient.runSqlHive(
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: Do we need to use `runSqlHive`?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for keeping working on this. Sorry for your long waiting. At this time, I hope we can merge your PR.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 7 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: sure, I will change to sql.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 3 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: changed

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 4 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: the whole exception message is 
`Not a file: file:/Users/qianyangyu/IdeaProjects/spark/target/tmp/spark-abc8c1ad-4a3a-420f-b4fc-58d995be9bb0/l1`, I will check the first part `Not a file:`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 5 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Changed.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 6 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: added the checking

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 8]
```diff
@@ -222,4 +223,127 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                 |  c1 int,
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}"}'""".stripMargin
+            sql(topDirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("SELECT * FROM tbl1"), Nil)
+            } else {
+              val msg = intercept[IOException] {sql("SELECT * FROM tbl1").show()
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: `sql("SELECT * FROM tbl1").show()` seems to be in the next line.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 9]
```diff
@@ -222,4 +223,127 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+              }.getMessage
+                assert(msg.contains("Not a file:"))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: indentation?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 10]
```diff
@@ -222,4 +223,127 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                 |CREATE EXTERNAL TABLE tbl2(
+                 |  c1 int,
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}/l1/"}'""".stripMargin
+            sql(l1DirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("SELECT * FROM tbl2"),
+                (1 to 2).map(i => Row(i, i, s"parq$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Can we merge 269 and 270 into one line here?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 11]
```diff
@@ -222,4 +223,127 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${s"${path.getCanonicalPath}/l1/l2/"}'""".stripMargin
+            sql(l2DirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("SELECT * FROM tbl3"),
+                (3 to 4).map(i => Row(i, i, s"parq$i")))
+            } else {
+              val msg = intercept[IOException] {sql("SELECT * FROM tbl3").show()
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: `sql("SELECT * FROM tbl3").show()`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 12]
```diff
@@ -222,4 +223,127 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
...
+                 |  c2 int,
+                 |  c3 string)
+                 |STORED AS parquet
+                 |LOCATION '${new File(s"${path}/l1/*").toURI}'""".stripMargin
+            sql(wildcardL1DirStatement)
+            if (parquetConversion == "true") {
+              checkAnswer(sql("SELECT * FROM tbl5"),
+                (1 to 4).map(i => Row(i, i, s"parq$i")))
+            } else {
+              val msg = intercept[IOException] {sql("SELECT * FROM tbl5").show()
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 13]
```diff
@@ -170,4 +170,156 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+
+  test("SPARK-25993 CREATE EXTERNAL TABLE with subdirectories") {
+    Seq(true, false).foreach { convertMetastore =>
+      withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> s"$convertMetastore") {
+        withTempDir { dir =>
+          try {
+            sql("USE default")
+            sql(
+              """
+                |CREATE EXTERNAL TABLE hive_orc(
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: I'm a little confused here.
@kevinyu98 . Do you want to get a table created by Hive here?
Usually, we use the table name, `hive_orc`, for that table. Please see https://github.com/apache/spark/pull/27130/files#diff-a8c26a35def87a13e6b59db19d9fb8a1R68 .

And, you still using `hiveClient.runSqlHive` at line 192. I'm wondering what is the test target in this PR~.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 8 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: thanks, changed.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 9 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 10 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 11 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: moved to next line.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 12 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 13 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: @dongjoon-hyun Thanks for pointing out this. I was using other test cases without thinking too much. I have changed the name. I also replaced the hiveClient.runSqlHive for the insert stmt.

**Comment by kevinyu98** [comment]
@dongjoon-hyun can we re-test it? I ran `HiveParquetSourceSuite` and `HiveOrcSourceSuite` locally, it works fine. thanks.

**Comment by dilipbiswal** [comment]
retest this please

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @kevinyu98 .
I agree with `HiveOrcSourceSuite.scala`. Could you update `HiveParquetSourceSuite.scala` like `HiveOrcSourceSuite.scala`? There is no need to be different.

**Comment by kevinyu98** [comment]
@dongjoon-hyun sure, thanks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 14]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+                   |  c1 int,
+                   |  c2 int,
+                   |  c3 string)
+                   |STORED AS orc
+                   |LOCATION '${s"${dir.getCanonicalPath}/l1/"}'""".stripMargin
+              sql(l1DirStatement)
+              val l1DirSqlStatement = s"SELECT * FROM tbl2"
+              if (convertMetastore) {
+                checkAnswer(sql(l1DirSqlStatement),
+                  (1 to 2).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: Let's make line 250 and 251 as one line like `HiveParquetSourceSuite`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 15]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+              } else {
+                checkAnswer(sql(l1DirSqlStatement),
+                  (1 to 6).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 16]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+                   |  c1 int,
+                   |  c2 int,
+                   |  c3 string)
+                   |STORED AS orc
+                   |LOCATION '${s"${dir.getCanonicalPath}/l1/l2/"}'""".stripMargin
+              sql(l2DirStatement)
+              val l2DirSqlStatement = s"SELECT * FROM tbl3"
+              if (convertMetastore) {
+                checkAnswer(sql(l2DirSqlStatement),
+                  (3 to 4).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 17]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+              } else {
+                checkAnswer(sql(l2DirSqlStatement),
+                  (3 to 6).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 18]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+                   |  c1 int,
+                   |  c2 int,
+                   |  c3 string)
+                   |STORED AS orc
+                   |LOCATION '${new File(s"${dir}/*").toURI}'""".stripMargin
+              sql(wildcardTopDirStatement)
+              val wildcardTopDirSqlStatement = s"SELECT * FROM tbl4"
+              if (convertMetastore) {
+                checkAnswer(sql(wildcardTopDirSqlStatement),
+                  (1 to 2).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 19]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+                   |  c1 int,
+                   |  c2 int,
+                   |  c3 string)
+                   |STORED AS orc
+                   |LOCATION '${new File(s"${dir}/l1/*").toURI}'""".stripMargin
+              sql(wildcardL1DirStatement)
+              val wildcardL1DirSqlStatement = s"SELECT * FROM tbl5"
+              if (convertMetastore) {
+                checkAnswer(sql(wildcardL1DirSqlStatement),
+                  (1 to 4).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 20]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+                   |  c1 int,
+                   |  c2 int,
+                   |  c3 string)
+                   |STORED AS orc
+                   |LOCATION '${new File(s"${dir}/l1/l2/*").toURI}'""".stripMargin
+              sql(wildcardL2Statement)
+              val wildcardL2SqlStatement = s"SELECT * FROM tbl6"
+              if (convertMetastore) {
+                checkAnswer(sql(wildcardL2SqlStatement),
+                  (3 to 6).map(i => Row(i, i, s"orc$i")))
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: ditto.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 21]
```diff
@@ -170,4 +170,166 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
...
+              } else {
+                checkAnswer(sql(wildcardL2SqlStatement), Nil)
+              }
+            }
+          } finally {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: Shall we convert this `try..finally` with `withTable` like `HiveParquetSourceSuite`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 14 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: sure

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 15 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 17 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 18 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 19 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: 👍

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 21 above)
Comment (kevinyu98) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala]: 👍

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @kevinyu98 .
Merged to master.

**Pull request closed without merging**